File: code/matlab/chapter5/shaded_band.m
function h = shaded_band(x, lo, hi, shade)
% shaded_band.m
% Plot a grayscale shaded credible band between pointwise lower and
% upper bounds. Useful for visualizing posterior credible intervals
% for time-varying quantities.
%
% Inputs:
%   x    : T-by-1 vector (numeric or datetime)
%   lo   : T-by-1 vector, lower bound
%   hi   : T-by-1 vector, upper bound
%   shade: grayscale level in [0,1] (optional, default = 0.85)
%
% Output:
%   h    : handle to the fill object

if nargin < 4
    shade = 0.85;
end

% ensure column vectors
x  = x(:);
lo = lo(:);
hi = hi(:);

% plot shaded region
h = fill([x; flipud(x)], [lo; flipud(hi)], ...
         shade*[1 1 1], 'EdgeColor','none');
end

File: code/R/chapter03/shaded_band.R
# shaded_band.R
# Plot a grayscale shaded credible band between pointwise lower and
# upper bounds. Useful for visualizing posterior credible intervals
# for time-varying quantities.
#
# Inputs:
#   x    : length-T vector (numeric or Date)
#   lo   : length-T vector, lower bound
#   hi   : length-T vector, upper bound
#   shade: grayscale level in [0,1] (optional, default = 0.85)
#
# Output:
#   NULL (the band is added to the current plot)
#
# Unlike MATLAB's fill, R's polygon draws on an existing plot region, so
# set up the axes first (e.g. plot(..., type = "n")) and call this before
# the lines that go on top of the band.

shaded_band <- function(x, lo, hi, shade = 0.85) {
    # ensure plain numeric vectors (Dates are drawn on their numeric scale)
    x  <- as.numeric(x)
    lo <- as.numeric(lo)
    hi <- as.numeric(hi)

    # plot shaded region
    polygon(c(x, rev(x)), c(lo, rev(hi)),
            col = gray(shade), border = NA)
}
